src/waveform.ts

```diff
@@ -140,23 +134,17 @@ export class APWaveForm extends Gtk.DrawingArea {
 
     const leftColor = this.safeLookupColor("accent_color");
 
-    // Because the cairo module isn't real, we have to use these to ignore `any`.
-    // We keep them to the minimum possible scope to catch real errors.
-    /* eslint-disable @typescript-eslint/no-unsafe-call */
-    /* eslint-disable @typescript-eslint/no-unsafe-member-access */
-    ctx.setLineCap(Cairo.LineCap.SQUARE);
-    ctx.setAntialias(Cairo.Antialias.NONE);
-    ctx.setLineWidth(2);
-
-    this.setSourceRGBA(ctx, leftColor);
-
-    ctx.moveTo(horizCenter, vertiCenter - height);
-    ctx.lineTo(horizCenter, vertiCenter + height);
-    ctx.stroke();
+    // Clip the snapshot to the widget area.
```

Comment:
In general the more sensible approach is to draw inside this rectangle, this code implies that things will be drawn outside and can appear half-way offscreen.

Reply:
You're correct. The reason is that the peaks are drawn "scaled up" by another 70%

Usually the waveform peaks are short (50-70%) of the widget height. I make them a bit larger so that there's not much empty space.

Should I stop clipping the peaks or should I instead clip the waveform lines before they are given their size is larger than widget.height?

Reply:
I am not familiar with the code, ideally one does the math and draws them at a size based on `height` in such a way that even after adding the position it does not go beyond the given size.

Without thinking too much, I would set the rectangle with height `factor * height` (factor goes from 0 to 1) and a y position given by `(height - factor * height) / 2.0` so that it looks centered.

Note that you have not ensure everything passed to the snapshot API (e.g. values for a rectangle) is an integer if you want to avoid blurryness. (just use a floor() fn whenever you do a division or multiple by a random float like `factor`).